File: 03/one.lua
package.path = package.path .. ";../modules/?.lua"
local tools = require "tools"

diagnosis = tools.readlines("input.txt")

bit_len = #diagnosis[1]
bits = {}
for i = 1, bit_len do
	bits[i] = {[0] = 0, [1] = 0}
end

for _, line in pairs(diagnosis) do
	for i = 1, bit_len do
		bits[i][tonumber(line:sub(i, i))] = bits[i][tonumber(line:sub(i, i))]- -1
	end
end


gamma = 0
for i = 1, bit_len do
	mcb = bits[i][0] > bits[i][1] and 0 or 1
	gamma = gamma << 1 | mcb
end

epsilon = gamma ~ (2^bit_len - 1)

print(gamma, epsilon);

print("Gamma rate * epsilon rate = " .. gamma * epsilon)
assert(gamma * epsilon == 3959450)

File: 3DS/03/source/one.cpp
#include <stdlib.h>
#include <3ds.h>
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include "utils.h"

using namespace std;

int one()
{
	auto diagnosis = readlines("romfs:/input.txt");

	int bitlen = diagnosis[0].length();

	int gamma_rate = 0;
	for (int i = 0; i < bitlen; i++)
	{
		int bit = 0;
		for (auto line : diagnosis)
		{
			bit += line[i] == '1' ? 1 : -1;
		}
		gamma_rate <<= 1;
		gamma_rate |= bit < 0 ? 0 : 1;
	}

	int epsilon_rate = (~gamma_rate) & ((1 << bitlen) - 1);
	return gamma_rate * epsilon_rate;
}
